### ISSUE TITLE
Reimplement reverse mode of gather

### ISSUE THREAD

**Issue by gbaydin** [DiffSharp/DiffSharp#386]
I've been looking at potential performance botlenecks in the current tensor code. Most of these are mainly remnants of code that relied on for loops over some tensor dimensions before more efficient implementations of the same were implemented (e.g., before the introduction of broadcasting ops).

After addressing #279 with #385, the last remaining such operation seems to be the reverse mode of `.gather(dim:int, indices:Tensor)` which I believe can be implemented much better and faster. This issue is to note that we should do this.

https://github.com/DiffSharp/DiffSharp/blob/a72afa5e615e743c56b9ec85238df4f1fea0458c/src/DiffSharp.Core/Tensor.fs#L2951

**Comment by gbaydin** [comment]
This is fixed in #400.

Quite happy about it as it changed this

```fsharp
| GatherT(a,dim,indices) -> 
    // TODO: The following is a minimal correct implementation. Faster and more memory efficient implementations should be possible.
    let tflat = td.flatten()
    let iflat = indices.flatten()
    if a.derivative.dim = 0 then a.derivative <- a.zerosLike() + a.derivative
    for i=0 to tflat.nelement-1 do
        let mutable t = tflat.[i]
        for k=0 to a.dim-1 do
            t <- t.unsqueeze(0)
        let j = iflat.[i].toScalar() :?> int
        let loc = flatIndexToIndex a.shape i
        loc.[dim] <- j
        a.derivative <- a.derivative.addSlice(loc, t)
    push (check(a.zeroLike(), a) :: tt)
```

to this

```fsharp
| GatherT(a,dim,indices) -> push (check(td.scatter(dim, indices, a.shape), a) :: tt)
```

**Comment by dsyme** [comment]
Much better!!
